src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
         /// <inheritdoc />
         public async Task<string> SetupAsync(SetupContext context)
         {
+            ShellSettings shellSettings = null;
             var initialState = context.ShellSettings.State;
             try
             {
-                var executionId = await SetupInternalAsync(context);
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
 
-                if (context.Errors.Any())
+                (var executionId, shellSettings) = await SetupInternalAsync(context);
+
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
-                    await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                    await _shellHost.ReloadShellContextAsync(shellSettings, eventSource: false);
```

Comment:
This is wrong, here to reload the shell we should use `context.ShellSettings` on which we revert back the state to the initial state, this is important to keep the tenant state consistent. Note: The reload may retrieve an updated state if for example the tenant was setup by another instance, which may explain why it failed here on this instance.

We should not use the internal `shellSettings` that we created locally only in memory for the setup, here if there is an error it would be still in the `Initializing` state, so the reload will not work as expected. We can only use this `ShellSettings` after a successful setup including the call to `UpdateShellSettingsAsync()`, as you do below for the `Succeeded` event.

Reply:
@jtkech what is a bit confusing for me here is why would be then call `_shellHost.UpdateShellSettingsAsync(shellSettings.AsRunning())` on the shellSettings that is in memory but then never set `context.ShellSettings.AsRunning()`. Does this mean when we call `UpdateShellSettingsAsync` on the memory ShellSettings, it'll override by writing it to the ShellSettings files which makes `context.ShellSettings` irrelevant?